substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/PreMainSupport.java

```diff
@@ -0,0 +1,231 @@
+     * -XX-premain:[full.qualified.premain.class2]:[premain options] <br>
+     *
+     * @param args original arguments for premain and main
+     * @return arguments for main class
+     */
+    public String[] retrievePremainArgs(String[] args) {
+        List<String> mainArgs = new ArrayList<>();
+
+        for (String arg : args) {
+            if (arg.startsWith(PREMAIN_OPTION_PREFIX)) {
+                String premainOptionKeyValue = arg.substring(PREMAIN_OPTION_PREFIX.length());
+                String[] pair = premainOptionKeyValue.split(":");
+                if (pair.length == 2) {
+                    premainOptions.put(pair[0], pair[1]);
+                }
+            } else {
+                mainArgs.add(arg);
+            }
+        }
+        return mainArgs.toArray(new String[0]);
+    }
+
+    public void invokePremain() {
+        for (PremainMethod premainMethod : premainMethods) {
+            try {
+                Object[] args = premainMethod.args;
+                if (premainOptions.containsKey(premainMethod.className)) {
+                    args[0] = premainOptions.get(premainMethod.className);
+                }
+                // premain method must be static
+                premainMethod.method.invoke(null, args);
+            } catch (Throwable t) {
+                VMError.shouldNotReachHere("Fail to execute " + premainMethod.className + ".premain", t);
+            }
+        }
+    }
+
+    /**
+     * This class is a dummy implementation of {@link Instrumentation} interface. It serves as the
+     * registered premain method's second parameter. At native image runtime, no actual
+     * instrumentation work can do. So all the methods here are empty.
+     */
+    public static class NativeImageNoOpRuntimeInstrumentation implements Instrumentation {
+
+        @Override
+        public void addTransformer(ClassFileTransformer transformer, boolean canRetransform) {
+        }
+
+        @Override
+        public void addTransformer(ClassFileTransformer transformer) {
+        }
+
+        @Override
+        public boolean removeTransformer(ClassFileTransformer transformer) {
+            return false;
+        }
+
+        @Override
+        public boolean isRetransformClassesSupported() {
+            return false;
+        }
+
+        @Override
+        public void retransformClasses(Class<?>... classes) throws UnmodifiableClassException {
+        }
+
+        @Override
+        public boolean isRedefineClassesSupported() {
+            return false;
+        }
+
+        @Override
+        public void redefineClasses(ClassDefinition... definitions) throws ClassNotFoundException, UnmodifiableClassException {
+        }
+
+        @Override
+        public boolean isModifiableClass(Class<?> theClass) {
+            return false;
+        }
+
+        @Override
+        public Class<?>[] getAllLoadedClasses() {
+            return new Class<?>[0];
```

Comment:
I think we should discuss this. How is this method typically used? 

Should we maybe return all the classes that were reached? I think that we need an explanation for this feature.

Reply:
In observability scenario, user can call this method to collect all jars loaded by the application.
I think return the reached classes in native image here is reasonable.

Reply:
> Should we maybe return all the classes that were reached?

this sounds good if it's possible